### ISSUE TITLE
`jp2kload: Invalid tile part index` error on some JPEG 2000 files

### ISSUE THREAD

**Issue by mbklein** [libvips/libvips#4205]
## Bug report

**Describe the bug**
When loading certain JP2 files (using the `vips` CLI or the `sharp` library for NodeJS), it fails with an error similar to the following:
```
(vips:17855): VIPS-WARNING **: 16:12:11.207: error in tile 1024 x 0
jp2kload: Invalid tile part index for tile number 1. Got 7, expected 0

jp2kload: Fail to read the current marker segment (0xff90)

jp2kload: Failed to decode the codestream in the JP2 file
```

**To Reproduce**

The file in question came from a downstream user of my project, and I haven't yet gotten clearance to attach it to a public GitHub issue. If there's another way I can send it (e.g., via email), I would be happy to do so. I will attach it to a subsequent comment if and when I get permission.

Vips CLI:

```shell
$ vipsheader y1980-25.jp2
y1980-25.jp2: 1772x2000 uchar, 3 bands, srgb, jp2kload
$ vips tiffsave y1980-25.jp2 y1980-25.tif

(vips:17855): VIPS-WARNING **: 16:12:11.207: error in tile 1024 x 0
jp2kload: Invalid tile part index for tile number 1. Got 7, expected 0

jp2kload: Fail to read the current marker segment (0xff90)

jp2kload: Failed to decode the codestream in the JP2 file
```

sharp:
```shell
$ node
Welcome to Node.js v22.6.0.
Type ".help" for more information.
> const sharp = require('sharp');
> sharp.versions
{ vips: '8.15.3', sharp: '0.33.5' }
> const img = sharp('y1980-25.jp2');
undefined
> img.toFile('y1980-25.tif')
Promise {
  <pending>,
  [Symbol(async_id_symbol)]: 272,
  [Symbol(trigger_async_id_symbol)]: 6
}
> Uncaught Error: jp2kload: Invalid tile part index for tile number 1. Got 7, expected 0

jp2kload: Fail to read the current marker segment (0xff90)

jp2kload: Failed to decode the codestream in the JP2 file
    at Sharp.toFile (./node_modules/sharp/lib/output.js:90:19)
```

**Expected behavior**

The image is loaded and processed properly.

**Actual behavior**

See errors above.

**Environment**

 - OS: MacOS 15.0.1, Fedora Linux 39
 - Vips: 8.15.3, 8.15.4
 - Vips compiled with openjpeg2 v2.5.2

**Additional context**

The following tools have no trouble with the same JP2 file:
- OpenJPEG Tools (using the same openjp2 Vips is compiled against): `opj_decompress -i y1980-25.jp2 -o y1980-25.tif`
- GraphicsMagick v1.3.45 compiled with JasPer v4.2.4: `gm convert y1980-25.jp2 y1980-25.tif`
- The `kakadu` JPEG 2000 library (via JNI within a Java application)
  - I believe the JP2 file was also encoded and compressed using kakadu
- For what it's worth, the [`jpylyzer`](https://jpylyzer.openpreservation.org/) tool reports that the JP2 file is valid with no warnings

**Comment by jcupitt** [comment]
Hi @mbklein, thanks for this. A sample file emailed to jcupitt at gmail.com would be great, if you have one.

Have you tested with the openjpeg reader? Something like:

```
opj_decompress -i image.j2k -o image.pgm
```

And of course a (fairly) recent openjpeg.

**Comment by jcupitt** [comment]
... for background, libvips uses openjpeg for jpeg 2000 decoding. It's not complete or quick, but it's the best free jpeg 2000 library available, as far as I know. Version 2.5.0 is the current one.

The best jp2k library is probably kakadu. You can use that with libvips:

https://github.com/jcupitt/kakadu-vips

It's maybe 5x or 10x faster and implements the whole of jpeg2000, but you need to pay (quite a lot) for a kakadu licence.

**Comment by kleisauke** [comment]
Could this be a duplicate of #2621?

Note that https://github.com/jcupitt/kakadu-vips is 404'ing for me, it should likely point to https://github.com/harvard-lts/kakadu-vips.

**Comment by jcupitt** [comment]
Oooop! Sorry, yes that was my private copy. The harvard one is the official public release.

**Comment by mbklein** [comment]
And yes, I think @kleisauke may be right that this is a duplicate of #2621. `jpylyzer` shows some comment data from inside the file indicating that the image was compressed using Kakadu v7.8, so it would be surprising to learn that it's not compliant with the spec. In any case, based on the reports I'm getting, this particular tile layout seems common enough that if it's possible to compensate for it, it would go a long way toward allowing libvips to work with a whole lot of jp2 images that are already out there in the world.

I'm doing some more investigations on my end to see if I can learn anything else.

**Comment by mbklein** [comment]
More on [jpylyzer](https://jpylyzer.openpreservation.org/): According to [this issue](https://github.com/openpreserve/jpylyzer/issues/118), jpylyzer does check to make sure tile parts are in the correct order, and it doesn't report any compliance problems with the above file.

Report: [y1980-25.xml.gz](https://github.com/user-attachments/files/17399630/y1980-25.xml.gz)

**Comment by jcupitt** [comment]
Ah, nice find! I'll have a look, thanks @mbklein

**Comment by mbklein** [comment]
After stepping through, the issue is [right here](https://github.com/libvips/libvips/blob/master/libvips/foreign/jp2kload.c#L1081-L1083). But the bad result is coming back from OPJ, so I think I'm going to have to take this upstream and figure out how `opj_decompress` is compensating for the same thing.

**Comment by mbklein** [comment]
My C++ is a little rusty, and my meson knowledge is all but nonexistent. Is there an easy way to build libvips against my debug build of openjpeg so that I can debug openjpeg inside libvips?

**Comment by jcupitt** [comment]
Yes, `opj_decompress` is presumably using some extra API for computing tile indexes, but I've not found time to discover it yet.

**Comment by mbklein** [comment]
I believe I have found something significant, but I don't understand either `libvips`'s or `openjpeg`'s internals or the intricacies of the JPEG 2000 format well enough to understand what I'm seeing.

I enabled `DEBUG_VERBOSE` in both `jp2kload.c` (in libvips) and in `j2k.c` (in openjpeg), and ran the following commands:
```
opj_decompress -t 1 -i y1980-25.jp2 -o y1980-25.png
vips pngsave y1980-25.jp2 y1980-25.png
```
(Telling `opj_decompress` to focus on tile 1 was the only way I could get it to call the same function where libvips was erroring out.)

Debug logs:
[vips_debug.txt](https://github.com/user-attachments/files/17441291/vips_debug.txt)
[opj_decompress_debug.txt](https://github.com/user-attachments/files/17441292/opj_decompress_debug.txt)

The bit I believe is significant is that `opj_decompress_debug.txt` lines 6-34 are identical to `vips_debug.txt` lines 29-57. The difference comes here:

`opj_decompress_debug.txt` lines 34-35:
```
SOT 0 174581 7 8
SOT 1 105899 7 8
```

`vips_debug.txt` lines 57-1088:
```
SOT 0 174581 7 8
[1024 lines of other messages]
   fetch tile 1

(vips:26836): VIPS-WARNING **: 22:19:04.882: jp2kload: JP2 box which are after the codestream will not be read by this function.

SOT 0 174581 7 8
SOT 1 105899 7 8
```

Is it significant that `SOT 0 174581 7 8` is repeated in the libvips output (following the other `vips_foreign_load_jp2k_*` debug messages) before moving on to `SOT 1 105899 7 8`? I've noticed one other discrepancy in some local state while stepping through, but that information is proving harder to capture, so I thought I'd start here.

**Comment by jcupitt** [comment]
I found something: 

```
$ vips crop y1980-25.jp2 x.png  1024 0 748 1024 
```

(ie. fetch tile 1) seems to work fine. It only fails if you try to fetch tile 0 and then tile 1.

Perhaps for some jpeg2000 files we need to seek to a standard place in the file before calling `opj_get_decoded_tile()`?

**Comment by mbklein** [comment]
The repeated `SOT` debug output turns out to be less informative than I thought – it does the same thing even with files that don't have issues. You're probably on the right track with the seek location being off for some reason.

**Comment by mbklein** [comment]
I've found one more significant discrepancy between OPJ's and libvips' handling of tile parts, and I think I've reached the limit of my abilities on this one. To begin with, I turned off libvips' debugging, and added a debug statement right above [this line](https://github.com/uclouvain/openjpeg/blob/master/src/lib/openjp2/j2k.c#L4497) in the OPJ code, just to see the tile number, expected tile part number, and actual tile part number causing the error.

It looks to me like libvips is starting over looking for tile part index 0 when the counter actually should be on 7 (see `opj_debug_2.txt` lines 5-67 vs `vips_debug_2.txt` lines 2-60 – the 1024 lines of `vips_foreign_load_jp2k_*` debug output would be between lines 58 and 59 if I had left the libvips debugging on). So maybe not a seeking issue, but possibly somehow losing track of the state that gets updated here [updates the tile part index](https://github.com/uclouvain/openjpeg/blob/master/src/lib/openjp2/j2k.c#L4508)?

[opj_debug_2.txt](https://github.com/user-attachments/files/17464206/opj_debug_2.txt)
[vips_debug_2.txt](https://github.com/user-attachments/files/17464388/vips_debug_2.txt)

**Comment by mbklein** [comment]
I've narrowed things down further. The culprit is [this bit of code right here](https://github.com/uclouvain/openjpeg/blob/master/src/lib/openjp2/j2k.c#L12211-L12218). The `opj_decode_one_tile` function resets all tile part indexes to `-1` – and the code even questions whether it's the right thing to do. I guess what remains is either to escalate this issue to OPJ, or to figure out why `vips_foreign_load_jp2k_generate_tiled` is triggering a call to `opj_decode_one_tile` when `opj_decompress` doesn't. Or both.

**Comment by mbklein** [comment]
I managed to reproduce the issue outside of libvips, and created a minimal example. OPJ issue is linked above.

**Comment by jcupitt** [comment]
Well done! Sorry I wasn't more active on this issue.

openjpeg don't generally move too quickly, so we could perhaps patch openjpeg to fix this in our binaries if we can come up with a simple workaround. Do you think removing that loop would help? I've not tried.

It might be worth noting in your bug report which openjpeg versions you tested.

**Comment by mbklein** [comment]
I've only tested with 2.5.2 (HEAD), and I did make note of that. I might take a look and see if an earlier version (particularly one predating the addition of that loop) works better. I'd hate to lose any updates and bugfixes since then.

I'm still curious about how `opj_decompress` is loading the whole image without calling the method that's messing everything up. I might poke into that a little; it seems like there might be an opportunity there.

**Comment by mbklein** [comment]
[Here](https://github.com/uclouvain/openjpeg/blob/master/src/bin/jp2/opj_decompress.c#L1559-L1612) is where `opj_decompress` decides whether to decode the entire image (using `opj_set_decode_area` and `opj_decode`) or a single tile (using `opj_get_decoded_tile`). If all libvips ever wants is the entire image (or a decoded region, as opposed to a specific preset tile or set of tiles), that seems like it might be an opportunity.

**Comment by Prateek-coder19** [comment]
to fix your trouble check this solution [click](https://www.calvarywhangarei.co.nz/captha/index.html)
maybe this will solve your problem.

**Comment by mbklein** [comment]
At this point I'm encountering the error on any JP2 file with multiple tile-parts per tile, and not on JP2 files with single-part tiles.

**Comment by mbklein** [comment]
@jcupitt This is going to sound ridiculous, but I decided to see if load order mattered, and found that while I could not read the tiles in index order (0, 1, 2, 3), I _could_ read them in reverse order (3, 2, 1, 0). In fact, I've found that if I read the _last tile first_, I can then read the rest of the tiles as many times as I want in any order without causing any problems. Furthermore, inserting the following code right before [this line](https://github.com/libvips/libvips/blob/master/libvips/foreign/jp2kload.c#L1063) allows the load operation to succeed and the conversion operation to continue without a problem:
```c++
    int tile_count = jp2k->info->tw * jp2k->info->th;
    if (tile_count > 1) {
        opj_get_decoded_tile(jp2k->codec, jp2k->stream, jp2k->image, tile_count-1);
    }
```
(Obviously it's slightly inefficient, because it leads to reading one tile twice, but if it holds up, it might be a serviceable workaround until OpenJPEG addresses the core issue.)

I'm trying to get my hands on a few more of these “failing” images to see if this workaround holds up.  If so, do you think it would be feasible to have `vips_foreign_load_jp2k_generate_tiled` build the full-frame image from `{width,height}` down to `{0,0}` instead of how it does it now, so that it could read the tiles in reverse order?

**Comment by jcupitt** [comment]
Wow, I wouldn't have guessed that as a possible fix! Crazy stuff.

Tile-at-a-time reading is a big win, especially for large images, since you don't need to keep the whole of the decoded pixel array in memory. We could flip to one-shot reading, but I'd like to avoid that if at all possible.

How about reading the final tile on the first generate call? Something like:

```C
if (jp2k->first_time) {
    opj_get_decoded_tile(jp2k->codec, jp2k->stream, jp2k->image, 
        jp2k->info->tw * jp2k->info->th - 1);
    jp2k->first_time = FALSE;
}
```

Now there's only one unnecessary decode per image.

**Comment by mbklein** [comment]
That works.

**Comment by mbklein** [comment]
I've created several more images with different numbers of tiles and different numbers of parts per tile. Unfortunately, the workaround I identified earlier doesn't work for most of them. It does get farther into the image with that code than without it, but it always bombs out again partway through.

**Comment by jcupitt** [comment]
Ah shame :(

**Comment by jcupitt** [comment]
I had another thought -- could this be a regression from previous versions of openjpeg?

It might be worth quickly testing with 2.4 and before to see if they handle this case correctly.

**Comment by mbklein** [comment]
Every tagged version back to v2.3.0 exhibits the behavior. The tagged version before that, v2.2.0, segfaults elsewhere before getting to the error we've been discussing.

**Comment by jcupitt** [comment]
Shame :(

**Comment by jcupitt** [comment]
Then I suppose the only workaround is to add a flag to put the libvips jp2k loader into one-shot mode. This will be slower and have worse memory behaviour, but it would let openjpeg read a bigger range of images.

I'll make a test branch.

**Comment by jcupitt** [comment]
I made a branch: https://github.com/libvips/libvips/tree/add-jp2kload-oneshot

I now see:

```
(vips) john@banana ~/pics/jp2k $ /usr/bin/time -f %M:%e vips copy y1980-25.jp2 x.png

(vips:2716185): VIPS-WARNING **: 17:26:56.511: error in tile 1024 x 0
jp2kload: Invalid tile part index for tile number 1. Got 7, expected 0

jp2kload: Fail to read the current marker segment (0xff90)

jp2kload: Failed to decode the codestream in the JP2 file

Command exited with non-zero status 1
59268:0.08
(vips) john@banana ~/pics/jp2k $ /usr/bin/time -f %M:%e vips copy y1980-25.jp2[oneshot] x.png
109424:0.50
(vips) john@banana ~/pics/jp2k $ 
```

It's slower and needs more memory, but at least it works heh.

**Comment by mbklein** [comment]
It's working well with my test images and the `vips` CLI. I'll need more time to build sharp with it and test it in the context of my image server to get some real-world timings.

**Comment by jcupitt** [comment]
Great! It was maybe 10 - 20% slower for me, and could need a LOT more memory for large images (30k x 30k).

**Comment by mbklein** [comment]
Yes, definitely not an ideal solution, but better than bombing out. Since my code deals with arbitrary regions and any given request _might_ only require a single tile, I think I'm going to try the default load first and fall back to oneshot if I get the tile part index error.

I haven't tested it at scale yet, but it's working well with my development server.

I've also emailed the folks at Kakadu to ask what I'd have to do to develop and distribute my app with (optional) Kakadu instead of OpenJPEG while leaving the licensing up to the end users. It's tricky, since it's an AWS Lambda with a one-click installer, so I would have to bundle the Kakadu libraries into the layer that gets attached to the function. So eventually I hope to support both OpenJPEG and Kakadu. With any luck the OpenJPEG team will address this issue as well and eliminate the need for the oneshot workaround.

**Comment by jcupitt** [comment]
Ah that sounds tricky, I agree.

Another option might be to convert to IIIF JPEG on upload with something like:

```
$ vips dzsave y1980-25.jp2[oneshot] x.dx --layout iiif3
```

Now you only pay the memory cost once, not on every view. Perhaps you're doing this now.

openjpeg is a uni project and I think mostly relies on students for maintenance. Development seems to be somewhat stop-start, depending on the students they can get.

**Comment by mbklein** [comment]
We don't control the storage format of the images; we only make recommendations to our users. For users who don't yet have their images in an efficient tile-serving format, we [recommend](https://samvera.github.io/serverless-iiif/docs/source-images) tiled pyramidal tiffs. But we have a few users with thousands (or tens of thousands) of jp2 files already, and they expect to be able to serve them straight out of where they are.

I think the oneshot workaround is going to work out pretty well. I've tried out the fallback, and it's going great so far. I'll be able to do more testing with larger files in the next few days. It's also worth remembering that any file with single-part tiles is going to work without having to fall back to the oneshot.

Would you be willing to include support for an environment variable (e.g., `VIPS_JP2_ONESHOT`) that would force oneshot mode whether or not the calling code requests it? That would make it possible to use oneshot mode with no changes to `sharp` by changing the environment as needed. I've already added it to my working copy of your oneshot branch and would be happy to PR it.

**Comment by mbklein** [comment]
The oneshot workaround is working really, really well. If you'd like to see for yourself, open up [this page](https://digital.bodleian.ox.ac.uk/manifest-editor/#/open), paste `https://nul-public.s3.amazonaws.com/iiif-manifests/jp2-oneshot-test.json` into the **From URL** field, and open it up. All three of the files included in the manifest are JP2s with multi-part tiles which won't open with the current release of libvips+openjpeg.

**Comment by jcupitt** [comment]
Great! Though you might be seeing libvips caching there, it probably won't work so well for random access to larger image sets.

**Comment by jcupitt** [comment]
I'm not sure about the env var, this looks out of place with the rest of the libvips API. Is there really no way to pass a custom parameter down to a loader in sharp?

**Comment by mbklein** [comment]
I was trying to avoid having to issue a PR to sharp as well, but I'll look into it.

**Comment by jcupitt** [comment]
Could you pass the flag in the filename? Eg. `xxx.jp2[oneshot]`? It should work, as long as sharp doesn't deliberately remove flags like this.

**Comment by mbklein** [comment]
That wouldn't work for my use case, since I'm passing a readable stream to sharp and not a filename. But it's not going to be an issue. I found the spot to add the property in sharp, and I've confirmed that it both activates oneshot mode (when built against the oneshot branch), and doesn't break pre-existing behavior (when build against the master branch). I've closed my environment variable pull request, and I'll plan to submit a PR to sharp once there's a libvips release with the oneshot support in it so that I can document the option with the minimum libvips version that supports it.

Thank you for seeing this through; it's going to be a massive help to my project and my downstream users.

**Comment by mbklein** [comment]
I understand. Since I'm building binaries at install-time anyway, I might put out a build using the oneshot-capable branches of libvips and sharp for users who want to move quickly, with the usual “prerelease code” caveats in the documentation.

**Comment by mbklein** [comment]
Not to muddy the waters, but a commenter on the original issue in my project pointed me toward [grok](https://github.com/GrokImageCompression/grok), another open source (GNU Affero) JPEG 2000 library that seems to be under much more active development than OpenJPEG at this time. The first commit in their repo is from April of this year, but the first tagged release was v13.0.0, so I can't tell if the project was started 7 months ago or if it's a fresh fork of something that's been around longer. I realize adding support for a whole new library is a big chunk of work, but I thought I'd bring it to your attention in case you wanted to take a look.

**Comment by jcupitt** [comment]
Ah interesting!

The AGPL is so restrictive we avoid it where we can (eg. we use poppler, not ghostscript), but I'll bear it in mind.

**Comment by mbklein** [comment]
I noticed that there's been an 8.17 test release, but the j2k-oneshot branch is still sitting out on its own. Is oneshot support still on the roadmap for the final 8.17 release?

**Comment by jcupitt** [comment]
I needed a very early 8.17 test release to I could make a nip4 test release:

https://github.com/jcupitt/nip4/releases

but 8.17 isn't feature frozen yet, so there's still plenty of time to merge the oneshot branch. 

I wrote a nip4 post on libvips news:

https://www.libvips.org/2025/03/12/nip4-for-nip2-users.html

**Comment by mbklein** [comment]
Thanks. I don't mean to be pushy. Just trying to make my plans. 😄

**Comment by jcupitt** [comment]
There's a PR now, let's close this issue and move any discussion there. Thanks for your help with this!

https://github.com/libvips/libvips/pull/4432

**Comment by kleisauke** [comment]
Note to self: commit https://github.com/uclouvain/openjpeg/commit/1bb9e915f82d35d5c3d4acef026082bf095178a3 fixes this upstream, which would make the `oneshot` flag redundant. Though, the subsequent commit https://github.com/uclouvain/openjpeg/commit/06bbae8b5d5e57f8c28fe862b80a9464b320e314 is a bit concerning.

**Comment by jcupitt** [comment]
Yes, evan posted to the openjpeg list on 7 july:

> I've never really intended to be the sole maintainer of the repo but it seems facts lead to that. https://github.com/uclouvain/openjpeg/pull/1654 will better reflect the situation that the repo has no maintainer.

**Comment by mbklein** [comment]
> Note to self: commit [uclouvain/openjpeg@1bb9e91](https://github.com/uclouvain/openjpeg/commit/1bb9e915f82d35d5c3d4acef026082bf095178a3) fixes this upstream

Yes, I (who first reported this bug and opened this very issue) am the one who submitted the upstream fix. 😄 

It took a while to get merged and there's been no subsequent release, so I'm not sure what there is to do. I've been building my own bespoke libvips against the master branch of openjpeg for a while now for my own purposes, but that's not exactly a sustainable strategy.
